File: demo/go/openapi_demo.go
package main

import (
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	PublisherKey = "your publisher key"
	ContentType  = "application/json"
)

func main() {
	// POST example:
	reqUrl := `https://openapi.toponad.com/v1/apps`
	reqBody := `{"limit":1}`
	respData, err := doRequest(http.MethodPost, reqUrl, reqBody)
	if err != nil {
		panic(err)
	}
	fmt.Println("post response: ", respData)

	// GET example:
	reqUrl = `https://openapi.toponad.com/v1/waterfall/units?placement_id='xxx'`
	respData, err = doRequest(http.MethodGet, reqUrl, "")
	if err != nil {
		panic(err)
	}
	fmt.Println("get response: ", respData)
}

// doRequest requests the given reqUrl and returns the response.
func doRequest(reqMethod, reqUrl, reqBody string) (string, error) {
	urlInfo, err := url.Parse(reqUrl)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest(reqMethod, reqUrl, bytes.NewReader([]byte(reqBody)))
	if err != nil {
		return "", err
	}

	// create the MD5 value of the request body.
	h := md5.New()
	h.Write([]byte(reqBody))
	contentMD5 := hex.EncodeToString(h.Sum(nil))
	contentMD5 = strings.ToUpper(contentMD5)
	// create the content of common headers except the header 'X-Up-Signature'.
	timestamp := strconv.Itoa(int(time.Now().Unix() * 1000))
	headerString := fmt.Sprintf("X-Up-Key:%s\nX-Up-Timestamp:%s", PublisherKey, timestamp)

	// create the final signature.
	sign := genSignature(reqMethod, contentMD5, ContentType, headerString, urlInfo.Path)
	req.Header.Set("Content-Type", ContentType)
	req.Header.Set("X-Up-Key", PublisherKey)
	req.Header.Set("X-Up-Timestamp", timestamp)
	req.Header.Set("X-Up-Signature", sign)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	respData, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(respData), nil
}

// genSignature returns the final signature
func genSignature(httpMethod, contentMD5, contentType, headerString, relativePath string) string {
	stringSection := []string{httpMethod, contentMD5, contentType, headerString, relativePath}
	stringToSign := strings.Join(stringSection, "\n")
	h := md5.New()
	h.Write([]byte(stringToSign))
	resultMD5 := hex.EncodeToString(h.Sum(nil))
	return strings.ToUpper(resultMD5)
}

File: demo/python/openapi_demo.py
import hashlib
import json
import urllib.parse

import requests
import time

PUBLISHER_KEY = 'your publisher key'
CONTENT_TYPE = "application/json"


def do_request(http_method, req_url, req_body):
    now_millis = int(time.time() * 1e3)

    # Create the final signature.
    content_md5 = hashlib.md5(req_body.encode()).hexdigest().upper()
    header_str = "X-Up-Key:{}\nX-Up-Timestamp:{}".format(PUBLISHER_KEY, str(now_millis))
    relative_path = urllib.parse.urlsplit(req_url).path
    sign_str = "{}\n{}\n{}\n{}\n{}".format(http_method, content_md5, CONTENT_TYPE, header_str, relative_path)
    final_sign = hashlib.md5(sign_str.encode()).hexdigest().upper()

    # do the request
    headers = {
        "X-Up-Key": PUBLISHER_KEY,
        "X-Up-Timestamp": str(now_millis),
        "X-Up-Signature": final_sign,
        "Content-Type": CONTENT_TYPE,
    }
    if http_method == "POST":
        response = requests.post(req_url, headers=headers, data=req_body)
    elif http_method == "GET":
        response = requests.get(req_url, headers=headers)
    else:
        # TODO
        response = "todo"
    return response.text


# POST example:
req_url = 'https://openapi.toponad.com/v1/apps'
req_body = {"limit": 1}
req_body_str = json.dumps(req_body, sort_keys=False)
response = do_request("POST", req_url, req_body=req_body_str)
print("post response: ", response)

# GET example:
req_url = "https://openapi.toponad.com/v1/waterfall/units?placement_id=xxx&is_abtest=0"
response = do_request("GET", req_url, req_body="")
print("get response: ", response)
